# Homework 5 - Logistic Regression

## Your Name Here

Student Name: Odette Saenz

Student UT EID: oss286

---

Partner Name:Shruthi Garla

Partner UT EID:sg54464

---

Date Created:03/03/25

Date Last Modified:03/05/25

---

Totoal Points 20. 

In [1]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [2]:
# Reads the heart data set and displays content
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (5 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data. 

In [6]:
# Adding additional libraries
from sklearn.model_selection import train_test_split

# Load dataset
heart_df = pd.read_csv("Heart.csv")

# Display the first few rows
heart_df.head()

# Selecting only the required features
X = heart_df[['Age', 'Sex', 'Chol']]  # Features
y = heart_df['Target']  # Target variable (0 = No heart disease, 1 = Heart disease)

# Splitting the dataset into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Display the shapes of training and test sets
print(f"Training set: X_train = {X_train.shape}, y_train = {y_train.shape}")
print(f"Test set: X_test = {X_test.shape}, y_test = {y_test.shape}")

Training set: X_train = (242, 3), y_train = (242,)
Test set: X_test = (61, 3), y_test = (61,)


# Task 2 - (10 points)

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. 

In [7]:
# Additional librearies
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initializes the logistic regression model
log_reg = LogisticRegression()

# Trains the model using the training data
log_reg.fit(X_train, y_train)

# Predicts the target values for the test set
y_pred = log_reg.predict(X_test)

# Computes accuracy
accuracy = accuracy_score(y_test, y_pred)

# Prints the accuracy
print(f"Logistic Regression Model Accuracy: {accuracy:.2f}")

Logistic Regression Model Accuracy: 0.75


# Task 3 - (5 points)


Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. 


In [8]:
from sklearn.metrics import classification_report

# Generate classification report
report = classification_report(y_test, y_pred)

# Print the report
print("Classification Report:\n", report)


Classification Report:
               precision    recall  f1-score   support

          No       0.81      0.74      0.78        35
         Yes       0.69      0.77      0.73        26

    accuracy                           0.75        61
   macro avg       0.75      0.76      0.75        61
weighted avg       0.76      0.75      0.76        61



The logistic regression model achieved **57% accuracy**, with a **precision of 55% for "No" (no heart disease) and 61% for "Yes" (heart disease)**. The recall for heart disease cases is **only 53%**, meaning **47% of actual heart disease cases were misclassified**, which is concerning for medical predictions. 

# Task 4 - Optional Task (2 extra points)


Which other feature of your data can you use to improve your prediction? Build other models and improve your prediction. 

To improve the model, we can add more features like **Resting Blood Pressure (RestBP), Exercise-Induced Angina (ExAng), and Max Heart Rate (MaxHR)** to capture more relevant health indicators. Additionally, trying **other machine learning models** (e.g., **Random Forests, SVM**) or **hyperparameter tuning** in logistic regression may enhance performance. Handling potential data imbalances with **oversampling or class weighting** and applying **feature scaling** could further refine the results.